src/vfio_device.rs

```diff
@@ -259,6 +273,36 @@ impl VfioContainer {
             .map_err(VfioError::KvmSetDeviceAttr)
     }
 
+    #[cfg(feature = "mshv")]
+    fn mshv_device_add_group(&self, group_fd: RawFd) -> Result<()> {
```

Comment:
Why is it explicitly named as "mshv_device_add_group()" and "kvm_device_add_group()"?
The function signature seems identical.

Reply:
Using `kvm_device_add_group` for MSHV is not correct and vice versa.

I can rename this function to `device_add_group`, but that will break existing clients because the old function is gone.

If you're okay with the downside, I can make the change.

Reply:
Shouldn't this function (and the kvm correspondent) need to be marked as `unsafe` because they're taking a `RawFd`? The caller needs to make sure that the fd is corresponding to the proper device.

I would be curious to know what happens when you pass an invalid fd. We could have a negative tests with this scenario (for example passing `-2` as the RawFd.

Reply:
Acutally, `kvm_device_add_group` is not public, so removing it should be fine.

@andreeaflorescu it is a private function. Also, the kernel will return an error if an invalid Fd is passed to it. I don't think it needs to be `unsafe`, but I don't have strong opinions here.

Reply:
In Rust functions that take as a parameter a `RawFd` are typically declared unsafe. I found the explanation here to be the most comprehensive in talking about why it needs to be declared as unsafe: https://github.com/rust-lang/rust/issues/72175#issuecomment-815433842.

Also, since it should be relatively straight forward to test the assumption that kvm returns an error I would suggest to add that as a test case. We've been adding negative tests to kvm-ioctls as well to test various KVM-related assumptions, and this has helped in identifying Linux versions that did not hold those assumptions (https://github.com/rust-vmm/kvm-ioctls/issues/152).

Reply:
Sure. I will have a look. Marking it as `unsafe` is easy enough. Let me check what needs to be done for adding some test cases.

Reply:
The crate hasn't been published yet, so it's acceptable to change APIs:)
@sboeuf what do you think?

Reply:
> In Rust functions that take as a parameter a RawFd are typically declared unsafe. I found the explanation here to be the most comprehensive in talking about why it needs to be declared as unsafe: rust-lang/rust#72175 (comment).

There is one simple way of dealing with this safety issue. `device_add_group` doesn't need to take `RawFd` at all. It should be about to take a reference to `VfioGroup` as the second parameter.